### PR TITLE
feat(validation): warn when KongIngress deprecated fields get populated

### LINKED ISSUE #3247: Admission webhook for `KongIngress` that warns about deprecation of certain fields
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

https://github.com/Kong/kubernetes-ingress-controller/issues/2941 marks some of the KongIngress fields as deprecated and scheduled for removal in 3.0 (https://github.com/Kong/kubernetes-ingress-controller/issues/3018).

It would be cool if `kubectl apply` of a KongIngress resource that uses the deprecated functionality showed a visual warning. This is a common behavior pattern in the Kubernetes API world.

### Proposed Solution

Implement an admission webhook for `KongIngress` that, if notices that a deprecated field of `KongIngress` is set, emits a warning that `kubectl` prints to stderr.

### Additional information

https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#response

> Admission webhooks can optionally return warning messages that are returned to the requesting client in HTTP Warning headers with a warning code of 299. Warnings can be sent with allowed or rejected admission responses.

### Acceptance Criteria

- [ ] As a `kubectl` user, if I `kubectl apply` a KongIngress that has a deprecated field set, I get a visual warning.

### PR DESCRIPTION
**Pull request by czeslavo**
**What this PR does / why we need it**:

- reorganize `admission.RequestHandler` to make it more readable (add a method per handled type, add response builder and move the handler to a separate file)
- add KongIngress validation returning warnings in case any of deprecated fields is populated 

**Which issue this PR fixes**:

Fixes #3247. 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/admission/handler.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+		h.Logger.WithError(err).Error("failed to marshal response")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		h.Logger.WithError(err).Error("failed to write response")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```
Comment (pmalek) [inline internal/admission/handler.go:68]: Ditto: 

```suggestion
	err := json.NewEncoder(w).Encode(review)
	if err != nil {
		h.Logger.WithError(err).Error("failed to encode response")
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
```

File: test/integration/kongingress_webhook_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	require.NoError(t, err)
+
+	t.Run("when deprecated fields are populated warnings are returned", func(t *testing.T) {
+		kongIngress := &configurationv1.KongIngress{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: "kong-ingress-validation-",
+			},
+			// Proxy field is deprecated, expecting warning for it.
+			Proxy: &configurationv1.KongIngressService{
+				Protocol: lo.ToPtr("tcp"),
```
Comment (pmalek) [inline test/integration/kongingress_webhook_test.go:62]: `address.Of` :) ?

https://github.com/Kong/kubernetes-ingress-controller/blob/ba97e9ee46d912eca610e8da23bb2ad89fc82100/internal/util/address/of.go#L3-L6

```suggestion
				Protocol: address.Of("tcp"),
```

File: internal/admission/handler.go [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+		if consumer.Username == oldConsumer.Username {
+			return responseBuilder.Allowed(true).Build(), nil
+		}
+		ok, message, err := h.Validator.ValidateConsumer(ctx, consumer)
+		if err != nil {
+			return nil, err
+		}
+		return responseBuilder.Allowed(ok).WithMessage(message).Build(), nil
+	default:
+		return nil, fmt.Errorf("unknown operation '%v'", string(request.Operation))
```
Comment (pmalek) [inline internal/admission/handler.go:173]: `%q` == `'%v'`

File: internal/admission/handler.go [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+func (h RequestHandler) handleKongConsumer(
+	ctx context.Context,
+	request admissionv1.AdmissionRequest,
+	responseBuilder *ResponseBuilder,
+) (*admissionv1.AdmissionResponse, error) {
+	consumer := configuration.KongConsumer{}
+	deserializer := codecs.UniversalDeserializer()
+	_, _, err := deserializer.Decode(request.Object.Raw,
+		nil, &consumer)
```
Comment (pmalek) [inline internal/admission/handler.go:144]: Aren't we a bit too conservative with line length :) ?

File: internal/admission/handler.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+		return
+	}
+	data, err := io.ReadAll(r.Body)
+	if err != nil {
+		h.Logger.WithError(err).Error("failed to read request from client")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	review := admissionv1.AdmissionReview{}
+	if err := json.Unmarshal(data, &review); err != nil {
```
[the remark this answers is not in the document]
Comment (czeslavo) [reply internal/admission/handler.go]: a51abda

File: internal/admission/handler.go (hunk 1 above)
Comment (czeslavo) [reply internal/admission/handler.go]: a51abda

File: test/integration/kongingress_webhook_test.go (hunk 2 above)
Comment (czeslavo) [reply test/integration/kongingress_webhook_test.go]: a51abda, although we could think about dropping `address.Of` - project already imports `github.com/samber/lo` which implements the same function. WDYT?

File: internal/admission/handler.go (hunk 3 above)
Comment (czeslavo) [reply internal/admission/handler.go]: a51abda

File: internal/admission/handler.go (hunk 4 above)
Comment (czeslavo) [reply internal/admission/handler.go]: Yes, we are 🙂: a51abda

File: test/integration/kongingress_webhook_test.go (hunk 2 above)
Comment (pmalek) [reply test/integration/kongingress_webhook_test.go]: 😬 I put sooooo much effort into this :)

I'm fine with this but we could make this properly e.g. blocking usages of `k8s.io/utils/pointer` and suggesting to use `lo.ToPtr()` instead. Might be done in a separate PR 👍

File: internal/admission/handler.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+		h.Logger.WithError(err).Error("failed to parse AdmissionReview object")
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	response, err := h.handleValidation(r.Context(), *review.Request)
+	if err != nil {
+		h.Logger.WithError(err).Error("failed to run validation")
```
[the remark this answers is not in the document]
Comment (pmalek) [reply internal/admission/handler.go:46]: Not sure what to make of this 🤔

**Review by pmalek — APPROVED** [review]

File: test/integration/kongingress_webhook_test.go (hunk 2 above)
Comment (czeslavo) [reply test/integration/kongingress_webhook_test.go]: > I'm fine with this but we could make this properly e.g. blocking usages of k8s.io/utils/pointer and suggesting to use lo.ToPtr() instead. Might be done in a separate PR 👍

Yeah, totally agree.

File: internal/admission/handler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (czeslavo) [reply internal/admission/handler.go:46]: The request is coming to us from the trusted Kubernetes API which makes this a false positive I think.

**Comment by pmalek** [comment]
I resolved all the conversations and it auto merged 🤷 🎉

**Pull request merged**